chain/client/src/stateless_validation/endorsement_tracker.rs

```diff
@@ -0,0 +1,141 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use near_chain_primitives::Error;
+use near_epoch_manager::EpochManagerAdapter;
+use near_primitives::block_body::ChunkEndorsementSignatures;
+use near_primitives::checked_feature;
+use near_primitives::sharding::{ChunkHash, ShardChunkHeader};
+use near_primitives::stateless_validation::ChunkEndorsement;
+use near_primitives::types::AccountId;
+
+use crate::Client;
+
+// This is the number of unique chunks for which we would track the chunk endorsements.
+// Ideally, we should not be processing more than num_shards chunks at a time.
+const NUM_CHUNK_ENDORSEMENTS_CACHE_COUNT: usize = 100;
+
+pub struct EndorsementTracker {
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    /// We store the validated chunk endorsements received from chunk validators
+    /// This is keyed on chunk_hash and account_id of validator to avoid duplicates.
+    /// Chunk endorsements would later be used as a part of block production.
+    chunk_endorsements: lru::LruCache<ChunkHash, HashMap<AccountId, ChunkEndorsement>>,
+}
+
+impl Client {
+    pub fn process_chunk_endorsement(
+        &mut self,
+        endorsement: ChunkEndorsement,
+    ) -> Result<(), Error> {
+        let chunk_header = self.chain.get_chunk(endorsement.chunk_hash())?.cloned_header();
+        self.chunk_endorsement_tracker.process_chunk_endorsement(chunk_header, endorsement)
```

Comment:
Tiny tiny amount of redirection here. Logic for `process_chunk_endorsement` pushed to `EndorsementTracker` from `Client`